torch/export/dynamic_shapes.py

```diff
@@ -1191,3 +1193,283 @@ def apply_fixes(path, dim, dummy):
         return dim  # unchanged dim
 
     return _tree_map_with_path(apply_fixes, dynamic_shapes, dynamic_shapes)
+
+
+def _postprocess_serialized_shapes(dynamic_shapes, dims):
+    """
+    Sorts dims and returns dictionary format
+    """
+    from torch.utils._sympy.numbers import int_oo
+
+    dims = {
+        k: {
+            "min": v["min"],
+            "max": None if v["max"] is int_oo else v["max"],
+            "derived": sorted(v["derived"]),
+        }
+        for k, v in sorted(dims.items())
+    }
+    return {
+        "dynamic_shapes": dynamic_shapes,
+        "dims": dims,
+    }
+
+
+def _dumps(dynamic_shapes, args, kwargs=None):
+    """
+    Utility function for dynamic shapes serialization, serializing a dynamic_shapes spec.
+    The serialized format is a dictionary containing 2 fields, "dynamic_shapes" and "dims".
+    Uses args & kwargs to distinguish between tensor-level and dim-level specs (only for Nones).
+
+    dynamic_shapes: A pytree structure mirroring the dynamic_shapes input to export():
+        - Each tensor input is represented with a list of values, non-tensor inputs with None.
+        - dynamic dimensions (i.e. symbols) in tensors and Dim enums are represented with strings.
+        - static dimensions are represented with ints.
+
+    dims: A dictionary mapping each symbol name to the min/max range and derived dim names.
+
+    For example:
+    ```
+    dx = Dim("dx", min=4, max=16)
+    dy = dx + 1
+
+    inputs = (
+        [
+            torch.randn(4, 4),
+            torch.randn(5, 4),
+        ],
+        torch.randn(4),
+        torch.randn(4, 4),
+        "hello",
+    )
+    dynamic_shapes = {
+        "a": [
+            (dx, 4),
+            (dy, 4),
+        ],
+        "b": (Dim.AUTO,),
+        "c": None,
+        "d": None,
+    }
+    out = _dumps(dynamic_shapes, inputs)
+    ```
+    would generate the following output:
+    ```
+    {
+        'dynamic_shapes': (
+            [
+                ['dx', 4],
+                ['dx + 1', 4],
+            ],
+            ['Dim.AUTO'],
+            [None, None],
+            None,
+        ),
+        'dims': {
+            'dx': {
+                'min': 4,
+                'max': 16,
+                'derived': ['dx + 1'],
+            },
+            'dy': {
```

Comment:
This should be better than just having dummy values.

I'm assuming you considered and rejected having the definition `dx + 1` in here? Why does that make things easier / harder?

Reply:
It's because when we construct DerivedDims, we're just applying a function on the root Dim, and min/max ranges aren't used - they're just derived from the root's properties.